### PR TITLE
Reduce string allocations in JAR classpath elements

### PR DESCRIPTION
**Pull request by retronym**
Add a one-element cache of the most recently encountered directory.

### CODE REVIEW

File: src/reflect/scala/reflect/io/ZipArchive.scala [hunk 1]
```diff
@@ -112,7 +112,7 @@ abstract class ZipArchive(override val file: JFile, release: Option[String]) ext
   private def ensureDir(dirs: java.util.Map[String, DirEntry], path: String, zipEntry: ZipEntry): DirEntry = {
     dirs get path match {
       case null =>
-        val parent = ensureDir(dirs, dirName(path), null)
+        val parent = ensureDir(dirs, dirNameUsingLast(path), null)
```
Comment (som-snytt) [inline src/reflect/scala/reflect/io/ZipArchive.scala]: Not sure this is useful. This is recursive, seeing successively shorter paths until root or some parent already is in dirs map. Also, it zaps the cached value for the first leaf, so second leaf recalculates; for getdir("/a/b"), getdir("/a/c").

**Review by som-snytt — APPROVED** [review]
Possible de-opt, but otherwise this code rang a bell.

It makes me wonder about efficiency trade-off of the volatile. Maybe ideally name-sharing would be indexes into the first leaf path. Also why isn't there built-in support for this common pattern? hashtag rhetorical question.

**Comment by retronym** [comment]
It would be handy to have a specialized hashmap that allowed use of a substring as a lookup key without actually needing to allocate the substring.

```scala
class StringHashMap[V] extends Map[String, V] {
   def getOrNull(string: S, prefixLength: Int): V = {
      val hash = improve(hashCodeOfSubstring(s, from , to))
      val bucket = entriesOrNil(table(hash))
      for (entry <- bucket) {
         val found = string.regionMatches(from, 0, entry.key, prefixLength)
         if (found) return entry.value
      }
      null
   }
}
```

The generalization might be:

```scala
trait KeyInjection[I, K] extends (I => K) {
   def hashOfK(i: I): Int
   def equalTo(i: I, k: K): Boolean
}

class InjectionLookupHashMap[K, V] {
  def getONullInjected[I](i: I)(impicit inj: KeyInjection[I, K]: B = {
  }
}
```

File: src/reflect/scala/reflect/io/ZipArchive.scala [hunk 2]
```diff
@@ -109,10 +109,10 @@ abstract class ZipArchive(override val file: JFile, release: Option[String]) ext
     }
   }
 
-  private def ensureDir(dirs: java.util.Map[String, DirEntry], path: String, zipEntry: ZipEntry): DirEntry = {
+  private def ensureDir(dirs: java.util.Map[String, DirEntry], path: String): DirEntry = {
```
Comment (som-snytt) [inline src/reflect/scala/reflect/io/ZipArchive.scala:112]: When I touched it on 2.13, I moved the method to local of getDir, cleaning up the signature.

Comment (retronym) [reply src/reflect/scala/reflect/io/ZipArchive.scala:112]: Thanks, that'll help my when merging next time.

**Comment by retronym** [comment]
In an odd coincidence I happened across https://github.com/openjdk/jdk/commit/fc728278c2bc4af041f2ffa1fba54bcdc4c0e4e3#diff-704561497d8c3df4c77bc8f35b15ad73 today which shows how `java.util.ZipFile` does a hash table lookup for `"foo"` _and_ `"foo/"` with minimum of overhead. That commit looked suspicious to me as names UTF-8 folder names can end in the `'/'.toByte`: https://twitter.com/retronym/status/1252134723431747584?s=20

**Pull request merged**